resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/StagedResourcesExpirationManager.scala

```diff
@@ -0,0 +1,121 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.deploy.rest.kubernetes
+
+import java.util.concurrent.{ScheduledExecutorService, TimeUnit}
+
+import scala.collection.JavaConverters._
+import scala.collection.concurrent.TrieMap
+import scala.collection.mutable
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.util.{Clock, Utils}
+
+private[spark] trait StagedResourcesExpirationManager {
+
+  def startMonitoringForExpiredResources(): Unit
+
+  def monitorResourceForExpiration(
+      resourceId: String,
+      podNamespace: String,
+      podLabels: Map[String, String],
+      kubernetesCredentials: PodMonitoringCredentials): Unit
+}
+
+private class StagedResourcesExpirationManagerImpl(
+      stagedResourcesStore: StagedResourcesStore,
+      kubernetesClientProvider: ResourceStagingServiceKubernetesClientProvider,
+      expirationExecutorService: ScheduledExecutorService,
+      clock: Clock,
+      expiredResourceTtlMs: Long,
+      resourceCleanupIntervalMs: Long)
+    extends StagedResourcesExpirationManager {
+
+  private val activeResources = TrieMap.empty[String, MonitoredResource]
+
+  override def startMonitoringForExpiredResources(): Unit = {
+    expirationExecutorService.scheduleAtFixedRate(
+        new CleanupRunnable(),
+        resourceCleanupIntervalMs,
+        resourceCleanupIntervalMs,
+        TimeUnit.MILLISECONDS)
+  }
+
+  override def monitorResourceForExpiration(
+      resourceId: String,
+      podNamespace: String,
+      podLabels: Map[String, String],
+      kubernetesCredentials: PodMonitoringCredentials): Unit = {
+    activeResources(resourceId) = MonitoredResource(
+        resourceId, podNamespace, podLabels, kubernetesCredentials)
+  }
+
+  private class CleanupRunnable extends Runnable with Logging {
+
+    private val expiringResources = mutable.Map.empty[String, ExpiringResource]
+
+    override def run(): Unit = {
+      // Make a copy so we can iterate through this while modifying activeResources
+      val activeResourcesCopy = Map.apply(activeResources.toSeq: _*)
+      for ((resourceId, resource) <- activeResourcesCopy) {
+        Utils.tryWithResource(
+            kubernetesClientProvider.getKubernetesClient(
+                resource.podNamespace, resource.kubernetesCredentials)) { kubernetesClient =>
+          val podsWithLabels = kubernetesClient
+            .pods()
+            .withLabels(resource.podLabels.asJava)
+            .list()
```

Comment:
Making this a watch is tricky because of the concurrency guarantees we have to satisfy here. Take a few scenarios, for example:

Suppose we had a policy that waits for pods to be deleted before starting to count down the expiration. Now suppose that before the watch is finished being established, the resources are deleted from the cluster. Then the watch never receives the delete event and thus never knows that resources should be cleaned up. In addition, watches don't give us the ability to know how many instances of a given resource are left, only that we have deleted a resource matching the label.

We could inverse this by saying that the watch will immediately begin counting down until expiration, and only when the watch receives an event like `ADDED` or `MODIFIED` then the expiration is halted. But in this case we have the inverse problem: if the pod is created before the watch is established, then we'll expire the resource prematurely.

One way we could make this all consistent is to enforce at the resource staging server that when resources are first sent to the cluster for pods with a given label set, that no pods with those labels are allowed to exist at the time that the resources are uploaded. Then we would open the watch immediately and begin ticking down expiration, canceling expiration if resources are added, and re-added expiration when resources are deleted. However again it's concerning that watches don't give us the ability to know how many of a given resource are remaining.

One other concern about watches is that they keep open connections to Kubernetes. We don't want to keep too many of these connections open at once.